packages/server/src/Connection.ts

```diff
@@ -216,21 +220,40 @@ export class Connection {
    * @public
    */
   public handleMessage(data: Uint8Array): void {
+    if (this.beforeHandleMessageQueue.length > 0 || this.beforeHandleMessageProcessing) {
+      this.beforeHandleMessageQueue.push(data)
+      return
+    }
+
+    this.processBeforeHandleMessage(data)
+  }
+
+  /** Process an incoming message */
+  private processBeforeHandleMessage(data: Uint8Array): void {
+    this.beforeHandleMessageProcessing = true
+
     const message = new IncomingMessage(data)
     const documentName = message.readVarString()
 
-    if (documentName !== this.document.name) return
+    if (documentName !== this.document.name) {
+      this.nextProcessBeforeHandleMessage()
+      return
+    }
 
     message.writeVarString(documentName)
 
     this.callbacks.beforeHandleMessage(this, data)
       .then(() => {
+        this.nextProcessBeforeHandleMessage()
```

Comment:
The ordering of things here seems a bit off to me: you trigger the processing of a next message *before* you've processed this one (below).

Reply:
The invocation of `this.nextProcessBeforeHandleMessage` will initiate a new async sequence and to me it seems possible that things go out-of-order again.